### ISSUE TITLE
StringIndexOutOfBoundsException: String index out of range: 8 when generating HTML report after update to 1.2.0

### ISSUE THREAD

**Issue by afaucogney** [arturbosch/detekt#2134]
The bug you're experiencing might have already be reported!
Please search in the [issues](https://github.com/arturbosch/detekt/issues) before creating one.

## Expected Behavior
Trying to upgrade to Detekt 1.20. Was in 1.0.0
HTML Report generation failed 

## Observed Behavior
Works with 1.10 and 1.1.1
Work if only XML and TXT generated

## Steps to Reproduce
Use my config I cannot give public

## Context
Use detekt-formatting

## Your Environment

* Version of detekt used: 1.2.0
* Version of Gradle used (if applicable): 5.4.1
* Operating System and version: Mac
* Link to your project (if it's a public repository):

```
Overall debt: 3d 11h 15min

Complexity Report:
        - 64695 lines of code (loc)
        - 47134 source lines of code (sloc)
        - 31244 logical lines of code (lloc)
        - 7087 comment lines of code (cloc)
        - 7202 McCabe complexity (mcc)
        - 566 number of total code smells
        - 15 % comment source ratio
        - 230 mcc per 1000 lloc
        - 18 code smells per 1000 lloc

Project Statistics:
        - number of properties: 4525
        - number of functions: 3968
        - number of classes: 2019
        - number of packages: 259
        - number of kt files: 1065

java.lang.StringIndexOutOfBoundsException: String index out of range: 8
        at java.lang.String.substring(String.java:1963)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.writeErrorLine(HtmlUtils.kt:38)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.access$writeErrorLine(HtmlUtils.kt:1)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1$1.invoke(HtmlUtils.kt:25)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1$1.invoke(HtmlUtils.kt)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_unionsKt.code(gen-tag-unions.kt:161)
        at kotlinx.html.Gen_tag_unionsKt.code$default(gen-tag-unions.kt:161)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1.invoke(HtmlUtils.kt:13)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1.invoke(HtmlUtils.kt)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_groupsKt.pre(gen-tag-groups.kt:131)
        at kotlinx.html.Gen_tag_groupsKt.pre$default(gen-tag-groups.kt:131)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.snippetCode(HtmlUtils.kt:12)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderFinding(HtmlOutputReport.kt:115)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.access$renderFinding(HtmlOutputReport.kt:36)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2$$special$$inlined$forEach$lambda$1.invoke(HtmlOutputReport.kt:95)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2$$special$$inlined$forEach$lambda$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tags_uKt.li(gen-tags-u.kt:20)
        at kotlinx.html.Gen_tags_uKt.li$default(gen-tags-u.kt:20)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2.invoke(HtmlOutputReport.kt:94)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_groupsKt.ul(gen-tag-groups.kt:143)
        at kotlinx.html.Gen_tag_groupsKt.ul$default(gen-tag-groups.kt:143)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1.invoke(HtmlOutputReport.kt:92)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_unionsKt.details(gen-tag-unions.kt:93)
        at kotlinx.html.Gen_tag_unionsKt.details$default(gen-tag-unions.kt:93)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderRule(HtmlOutputReport.kt:83)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderGroup(HtmlOutputReport.kt:78)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.access$renderGroup(HtmlOutputReport.kt:36)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderFindings$1.invoke(HtmlOutputReport.kt:68)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderFindings$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.ApiKt.visitAndFinalize(api.kt:93)
        at kotlinx.html.Gen_consumer_tagsKt.div(gen-consumer-tags.kt:186)
        at kotlinx.html.Gen_consumer_tagsKt.div$default(gen-consumer-tags.kt:186)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderFindings(HtmlOutputReport.kt:64)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.render(HtmlOutputReport.kt:46)
        at io.gitlab.arturbosch.detekt.api.OutputReport.write(OutputReport.kt:27)
        at io.gitlab.arturbosch.detekt.cli.OutputFacade.handleOutputReport(OutputFacade.kt:50)
        at io.gitlab.arturbosch.detekt.cli.OutputFacade.run(OutputFacade.kt:42)
        at io.gitlab.arturbosch.detekt.cli.runners.Runner.execute(Runner.kt:31)
        at io.gitlab.arturbosch.detekt.cli.Main.main(Main.kt:16)
```

**Comment by arturbosch** [comment]
@afaucogney thanks for reporting this. You mentioned context=detekt-formatting.
Does this crash only occur when adding the formatting plugin? Without formatting the rendering does it's job?
Can you debug which file and finding lead to this crash?

**Comment by afaucogney** [comment]
@arturbosch I tried to disable, and this is working, so it does mean that this is the formatting plugin that failed !

Here is my confirguration:
```
dependencies {
    // Detekt Ktlint
    //noinspection GradleDependency
    detekt "io.gitlab.arturbosch.detekt:detekt-cli:1.2.0"
    //noinspection GradleDependency
    detekt "io.gitlab.arturbosch.detekt:detekt-formatting:1.2.0"
}

// Add baseline when it's PR
def detektBaseline() {
    if (System.getenv('IsPR'))
        return null
    else
        return file("${rootProject.projectDir}/config/detekt/detekt-baseline-v90.xml")
}

// Settings
detekt {
    toolVersion = "1.2.0"
    config = files("${rootProject.projectDir}/config/detekt/detekt-config-ide.yml")
    buildUponDefaultConfig = false
    disableDefaultRuleSets = false
    input = files(
            "src/main/java",
            "src/developpement/java",
            "src/test/kotlin",
            "src/test/java",
            "build/generated/source/kapt"
    )
    baseline = detektBaseline()
    parallel = true
    debug = true
    ignoreFailures = false
    reports {
        xml {
            enabled = true
            // Enable/Disable XML report (default: true)
            destination = file("build/reports/detekt/detekt-report.xml")
            // Path where XML report will be stored (default: `build/reports/detekt/detekt.xml`)
        }
        txt {
            enabled = true
            destination = file("build/reports/detekt/detekt-report.txt")
        }
        html {
            enabled = true
            destination = file("build/reports/detekt/detekt-report.html")
        }

    }
}
```

Here is the config for formating:
```
formatting:
  active: true
  android: true
  autoCorrect: true
  ChainWrapping:
    active: true
    autoCorrect: true
  CommentSpacing:
    active: false
    autoCorrect: false
  Filename:
    active: true
  FinalNewline:
    active: false
    autoCorrect: true
  ImportOrdering:
    active: true
    autoCorrect: true
  Indentation:
    active: false
    autoCorrect: true
    indentSize: 4
    continuationIndentSize: 4
  MaximumLineLength:
    active: true
    maxLineLength: 130
  ModifierOrdering:
    active: true
    autoCorrect: true
  NoBlankLineBeforeRbrace:
    active: true
    autoCorrect: true
  NoConsecutiveBlankLines:
    active: true
    autoCorrect: true
  NoEmptyClassBody:
    active: true
    autoCorrect: true
  NoLineBreakAfterElse:
    active: true
    autoCorrect: true
  NoLineBreakBeforeAssignment:
    active: true
    autoCorrect: true
  NoMultipleSpaces:
    active: true
    autoCorrect: true
  NoSemicolons:
    active: true
    autoCorrect: true
  NoTrailingSpaces:
    active: false
    autoCorrect: true
  NoUnitReturn:
    active: true
    autoCorrect: true
  NoUnusedImports:
    active: true
    autoCorrect: true
  NoWildcardImports:
    active: true
    autoCorrect: true
  ParameterListWrapping:
    active: false
    autoCorrect: true
    indentSize: 4
  SpacingAroundColon:
    active: true
    autoCorrect: true
  SpacingAroundComma:
    active: true
    autoCorrect: true
  SpacingAroundCurly:
    active: true
    autoCorrect: true
  SpacingAroundKeyword:
    active: true
    autoCorrect: true
  SpacingAroundOperators:
    active: true
    autoCorrect: true
  SpacingAroundRangeOperator:
    active: true
    autoCorrect: true
  StringTemplate:
    active: true
    autoCorrect: true
```

However this will be hard to find the issue. But I can give you in private the xml files to run a deeper debug ride.

**Comment by arturbosch** [comment]
That's strange because we do not print the code location for the formatting rules.
![2019-11-28-105347_1211x621_scrot](https://user-images.githubusercontent.com/20924106/69796049-955ba300-11cd-11ea-89a5-5b22532b0923.png)

These two files can only be triggered for the standard rule set:
```
at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.snippetCode(HtmlUtils.kt:12)
at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderFinding(HtmlOutputReport.kt:
```

**Comment by afaucogney** [comment]
Anything I can do ? @arturbosch

**Comment by arturbosch** [comment]
Please download the 1.2.0 cli-all.jar from bintray and run detekt over project with `--report html:./detekt.html` and see if it still crashes for you. If not try downloading the formatting jar and add `--plugins formatting.jar` to the command.
I suspect that it already crashes with only the cli run. Thanks!

**Comment by dafi** [comment]
I've downloaded [cli 1.2.1](https://bintray.com/arturbosch/code-analysis/detekt/1.2.1#files/io%2Fgitlab%2Farturbosch%2Fdetekt%2Fdetekt-cli%2F1.2.1) and [formatting 1.2.1](https://bintray.com/arturbosch/code-analysis/detekt/1.2.1#files/io%2Fgitlab%2Farturbosch%2Fdetekt%2Fdetekt-formatting%2F1.2.1) but the the bug is yet present.

I fixed the source code and now the report is generated without any crash

The modified file is [HtmlUtils.kt](https://github.com/arturbosch/detekt/blob/master/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlUtils.kt#L24) I don't pull a request because I'm not sure the patched code is the real correct way to fix the bug

```kotlin
internal fun FlowContent.snippetCode(lines: Sequence<String>, location: SourceLocation, length: Int) {
    pre {
        code {
            val dropLineCount = max(location.line - 1 - EXTRA_LINES_IN_SNIPPET, 0)
            val takeLineCount = EXTRA_LINES_IN_SNIPPET + 1 + min(location.line - 1, EXTRA_LINES_IN_SNIPPET)
            var currentLineNumber = dropLineCount + 1
            var errorLength = length
            lines
                .drop(dropLineCount)
                .take(takeLineCount)
                .forEach { line ->
                    span("lineno") { text("%1$4s ".format(currentLineNumber)) }
                    if (currentLineNumber >= location.line && errorLength > 0) {
                        val column = if (currentLineNumber == location.line) location.column - 1 else 0
                        // dafi: The the report generation crash
                        errorLength -= if (column < line.length) {
                            writeErrorLine(line, column, errorLength) + 1 // we need to consume the \n
                            } else {
                                1
                            }
                    } else {
                        text(line)
                    }
                    text("\n")
                    currentLineNumber++
                }
        }
    }
}

```

**Comment by BraisGabin** [comment]
I don't have a computer near to check it but this seems a problem in a rule providing an incorrect location.

**Comment by schalkms** [comment]
@dafi thanks for investigating this. Would you mind creating a PR anyway? This helps us to get a feeling for the problem.

**Comment by dafi** [comment]
@schalkms ok I created the pull request #2160 but I'm pretty sure the fix must be done is the rule code as @BraisGabin says

**Comment by BraisGabin** [comment]
As a continuation of #2160

- We should add a test to avoid regressions, and
- We should probably add a message asking for report that error so we can fix the rules in first place. Otherwise we are displaying misleading snippets.

**Comment by eric-labelle** [comment]
If I got that right, the OutOfBoundException should have been fixed by #2160 which was released in 1.2.2 although I still have the exception with 1.2.2 as soon as I use `detekt-formatting:1.2.2`

```
java.lang.StringIndexOutOfBoundsException: String index out of range: 67
        at java.lang.String.substring(String.java:1963)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.writeErrorLine(HtmlUtils.kt:38)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.access$writeErrorLine(HtmlUtils.kt:1)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1$1.invoke(HtmlUtils.kt:25)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1$1.invoke(HtmlUtils.kt)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_unionsKt.code(gen-tag-unions.kt:161)
        at kotlinx.html.Gen_tag_unionsKt.code$default(gen-tag-unions.kt:161)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1.invoke(HtmlUtils.kt:13)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt$snippetCode$1.invoke(HtmlUtils.kt)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_groupsKt.pre(gen-tag-groups.kt:131)
        at kotlinx.html.Gen_tag_groupsKt.pre$default(gen-tag-groups.kt:131)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlUtilsKt.snippetCode(HtmlUtils.kt:12)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderFinding(HtmlOutputReport.kt:115)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.access$renderFinding(HtmlOutputReport.kt:36)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2$$special$$inlined$forEach$lambda$1.invoke(HtmlOutputReport.kt:95)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2$$special$$inlined$forEach$lambda$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tags_uKt.li(gen-tags-u.kt:20)
        at kotlinx.html.Gen_tags_uKt.li$default(gen-tags-u.kt:20)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2.invoke(HtmlOutputReport.kt:94)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1$2.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_groupsKt.ul(gen-tag-groups.kt:143)
        at kotlinx.html.Gen_tag_groupsKt.ul$default(gen-tag-groups.kt:143)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1.invoke(HtmlOutputReport.kt:92)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderRule$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.Gen_tag_unionsKt.details(gen-tag-unions.kt:93)
        at kotlinx.html.Gen_tag_unionsKt.details$default(gen-tag-unions.kt:93)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderRule(HtmlOutputReport.kt:83)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderGroup(HtmlOutputReport.kt:78)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.access$renderGroup(HtmlOutputReport.kt:36)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderFindings$1.invoke(HtmlOutputReport.kt:68)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport$renderFindings$1.invoke(HtmlOutputReport.kt:36)
        at kotlinx.html.ApiKt.visit(api.kt:80)
        at kotlinx.html.ApiKt.visitAndFinalize(api.kt:93)
        at kotlinx.html.Gen_consumer_tagsKt.div(gen-consumer-tags.kt:186)
        at kotlinx.html.Gen_consumer_tagsKt.div$default(gen-consumer-tags.kt:186)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.renderFindings(HtmlOutputReport.kt:64)
        at io.gitlab.arturbosch.detekt.cli.out.HtmlOutputReport.render(HtmlOutputReport.kt:46)
        at io.gitlab.arturbosch.detekt.api.OutputReport.write(OutputReport.kt:27)
        at io.gitlab.arturbosch.detekt.cli.OutputFacade.handleOutputReport(OutputFacade.kt:50)
        at io.gitlab.arturbosch.detekt.cli.OutputFacade.run(OutputFacade.kt:42)
        at io.gitlab.arturbosch.detekt.cli.runners.Runner.execute(Runner.kt:31)
        at io.gitlab.arturbosch.detekt.cli.Main.main(Main.kt:16)
```

**Comment by BraisGabin** [comment]
Do you have a way to reproduce the error, @eric-labelle?
